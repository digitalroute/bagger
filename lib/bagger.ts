import { BaggerResponse } from './response';
import { BaggerRequest, Method } from './request';
import { BaggerRequestBody } from './request_body';
import { BaggerConfiguration, BaggerConfigurationInternal } from './configuration';
import { OpenAPIObject, SecuritySchemeType } from 'openapi3-ts';
import { BaggerParameter, ParameterType } from './parameters';
import { BaggerComponentAdder, BaggerComponentGetter, BaggerSecurityComponent } from './component';
import { SchemaStorage, SchemaDefinition } from './schema_storage';

export class Bagger {
  private internalConfiguration = new BaggerConfigurationInternal();
  private schemaStorage = new SchemaStorage();
  private configuration: BaggerConfiguration;
  private componentAdder: BaggerComponentAdder;
  private componentGetter: BaggerComponentGetter;

  public constructor() {
    this.configuration = new BaggerConfiguration(this.internalConfiguration);
    this.componentAdder = new BaggerComponentAdder(this.internalConfiguration);
    this.componentGetter = new BaggerComponentGetter(this.internalConfiguration);
  }

  /**
   * Get the validation object for a specific path and method. The result can be used for validation.
   * This comes from `bagger.parameter()` and `bagger.requestBody()`, and they have to be declared on the request before getting them.
   * @param path The url/path to the request
   * @param method HTTP method like 'GET' or 'PUT'
   * @param contentType
   */
  public getRequestSchema(path: string, method: string, contentType: string = 'application/json'): SchemaDefinition {
    return this.schemaStorage.getRequestSchema(path, method, contentType);
  }

  /**
   * Creates a Request object
   * @param path Path representing the request
   * @param method HTTP method for this path. Multiple methods can be added for a path by calling
   * addRequest for the same path multiple times.
   * @example
   * ```js
   * const bagger = require('.');
   * const joi = require('joi');
   *
   * bagger
   *   .addRequest('/bags', 'get')
   *   .addTag('bags')
   *   .addTag('getters')
   *   .addSecurity('OAuth2')
   *   .addParameter(
   *     bagger
   *       .parameter()
   *       .query('bagSize')
   *       .schema(joi.string().valid('10L', '20L', '30L')
   *       .required(true)
   *   )
   *   .addResponse(
   *     bagger
   *       .response(204)
   *       .description('Fetched bag of given size')
   *   );
   * ```
   */
  public addRequest(path: string, method: Method): BaggerRequest {
    const request = new BaggerRequest(path, method, this.schemaStorage);
    this.internalConfiguration.addRequest(request);
    return request;
  }

  /**
   * Add a reusable component that can be referenced to.
   */
  public addComponent(): BaggerComponentAdder {
    return this.componentAdder;
  }

  /**
   * Get a reusable component.
   */
  public getComponent(): BaggerComponentGetter {
    return this.componentGetter;
  }

  /**
   * Get the configuration object for bagger. This defines the configuration for all endpoints.
   * @example
   * ```js
   * const bagger = require('bagger');
   * bagger
   *   .configure()
   *   .info({
   *     title: 'Bagger API',
   *     version: 'v1',
   *     description: 'Provides resources to building swagger definitions'
   *   })
   *   .addServer({
   *     url: 'https://localhost:3000',
   *     description: 'Local development'
   *   })
   * ...
   * ```
   */
  public configure(): BaggerConfiguration {
    return this.configuration;
  }

  /**
   * Compiles bagger into an OpenAPI schema. This has to be done _after_ all requests has been added to
   * bagger.
   * @returns An OpenAPI schema that can be used to render a UI.
   */
  public compile(): OpenAPIObject {
    return this.internalConfiguration.compile();
  }

  /**
   * Creates a Response object
   * @param httpCode The HTTP Code that the response represents
   * @returns A bagger response that can be used to create a compiled Swagger definition.
   * @example
   * ```js
   * const bagger = require('digitalroute/bagger');
   *
   * const getBags = bagger
   *   .response(200)
   *   .description('Successfully fetched all bags')
   *   .content('text/plain', { type: 'string' });
   * ```
   */
  public response(httpCode: number): BaggerResponse {
    return new BaggerResponse(httpCode);
  }

  public securityComponent(name: string, type: SecuritySchemeType): BaggerSecurityComponent {
    return new BaggerSecurityComponent(name, type);
  }

  /**
   * Create a request body used for defining a body in a bagger request
   * @example
   * ```js
   * const bagger = require('.');
   * const joi = require('joi');
   *
   * const body = bagger
   *   .requestBody()
   *   .description('Create a bag')
   *   .content('application/json',
   *     joi.object().keys({
   *       type: joi.string().valid('backpack', 'duffel', 'sports').required(),
   *       size: joi.array().items(joi.string().valid('10L', '20L', '30L').required()).required(),
   *       description: joi.string().optional()
   *     })
   *   )
   *   .required(true);
   *
   * bagger
   *   .addRequest('/bags', 'post')
   *   .body(body)
   * ...
   * ```
   */
  public requestBody(): BaggerRequestBody {
    return new BaggerRequestBody();
  }

  /**
   * Create a parameter used for defining query, path, cookie or header parameter in bagger requests.
   * @example
   * ```js
   * const bagger = require('bagger');
   * const parameter = bagger
   *   .parameter()
   *   .path('bagId')
   *   .schema(joi.string().required())
   *   .description('ID of one bag')
   *   .explode(true)
   *   .required(true);
   *
   * bagger
   *   .addRequest('/bags/{bagId}', 'get')
   *   .addParameter(parameter)
   * ...
   * ```
   */
  public parameter(): { [key in ParameterType]: (name: string) => BaggerParameter } {
    return {
      query: (name: string) => new BaggerParameter('query', name),
      path: (name: string) => new BaggerParameter('path', name),
      cookie: (name: string) => new BaggerParameter('cookie', name),
      header: (name: string) => new BaggerParameter('header', name)
    };
  }
}

export const defaultInstance = new Bagger();
