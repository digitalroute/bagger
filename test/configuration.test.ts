import bagger from '../lib/bagger';
import joi from 'joi';

describe('Bagger compiler', () => {
  const defaultInfo = {
    title: 'Bagger API',
    version: 'v1',
    description: 'Provides resources related to building swagger definitions'
  };

  test('An empty definition should compile', () => {
    bagger.configure().info(defaultInfo);
    expect(bagger.compile()).toMatchSnapshot();
  });

  test('A request can be included', () => {
    bagger.configure().info(defaultInfo);
    bagger
      .addRequest('/bags', 'get')
      .addTag('bags')
      .addTag('users')
      .addResponse(bagger.response(200).description('Good fetch'));
    expect(bagger.compile()).toMatchSnapshot();
  });

  test('Add a request and get the schema', () => {
    bagger.configure().info(defaultInfo);
    const schema = joi.object().keys({
      bagType: joi
        .string()
        .valid('backpack', 'sportsbag')
        .required()
    });
    bagger
      .addRequest('/bags', 'post')
      .addTag('bags')
      .body(bagger.requestBody().content('application/json', schema))
      .addResponse(bagger.response(200).description('Bought a bag!'));
    expect(bagger.getRequestSchema('/bags', 'post')).toEqual({ body: schema });
    expect(bagger.compile()).toMatchSnapshot();
  });

  test('Add security scheme and compile', () => {
    bagger.configure().info(defaultInfo);
    bagger.addComponent().security(
      bagger
        .securityComponent('ApiKeyAuth', 'apiKey')
        .in('header')
        .name('Authorization')
    );
    bagger
      .addRequest('/bags', 'get')
      .addTag('bags')
      .addResponse(bagger.response(200).description('Got bags!'));
    //expect(bagger.compile()).toMatchSnapshot();
  });
});
