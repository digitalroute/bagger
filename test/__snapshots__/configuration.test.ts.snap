// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bagger compiler A request can be included 1`] = `
Object {
  "components": Object {
    "schemas": Object {},
  },
  "info": Object {
    "description": "Provides resources related to building swagger definitions",
    "title": "Bagger API",
    "version": "v1",
  },
  "openapi": "3.0.0",
  "paths": Object {
    "/bags": Object {
      "get": Object {
        "responses": Object {
          "200": Object {
            "description": "Good fetch",
          },
        },
        "tags": Array [
          "bags",
          "users",
        ],
      },
    },
  },
}
`;

exports[`Bagger compiler Add a request and get the schema 1`] = `
Object {
  "components": Object {
    "schemas": Object {},
  },
  "info": Object {
    "description": "Provides resources related to building swagger definitions",
    "title": "Bagger API",
    "version": "v1",
  },
  "openapi": "3.0.0",
  "paths": Object {
    "/bags": Object {
      "get": Object {
        "responses": Object {
          "200": Object {
            "description": "Good fetch",
          },
        },
        "tags": Array [
          "bags",
          "users",
        ],
      },
      "post": Object {
        "requestBody": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "properties": Object {
                  "bagType": Object {
                    "enum": Array [
                      "backpack",
                      "sportsbag",
                    ],
                    "type": "string",
                  },
                },
                "required": Array [
                  "bagType",
                ],
                "type": "object",
              },
            },
          },
          "required": false,
        },
        "responses": Object {
          "200": Object {
            "description": "Bought a bag!",
          },
        },
        "tags": Array [
          "bags",
        ],
      },
    },
  },
}
`;

exports[`Bagger compiler An empty definition should compile 1`] = `
Object {
  "components": Object {
    "schemas": Object {},
  },
  "info": Object {
    "description": "Provides resources related to building swagger definitions",
    "title": "Bagger API",
    "version": "v1",
  },
  "openapi": "3.0.0",
  "paths": Object {},
}
`;
